src/ops_complex.rs

```diff
@@ -605,15 +600,25 @@ pub fn generate_permutation_matrix(
         32,
         stream,
     )?;
+    let (unique_variable_indexes, temp_storage) = temp_storage.split_at_mut(num_bytes);
+    let unique_variable_indexes = unsafe { unique_variable_indexes.transmute_mut() };
+    let (run_lengths, temp_storage) = temp_storage.split_at_mut(num_bytes);
+    let run_lengths = unsafe { run_lengths.transmute_mut() };
+    let (num_runs_out, encode_temp_storage) = temp_storage.split_at_mut(cell_size);
+    let num_runs_out = unsafe { &mut num_runs_out.transmute_mut()[0] };
     set_to_zero(run_lengths, stream)?;
     encode(
-        temp_storage,
+        encode_temp_storage,
         sorted_variable_indexes,
         unique_variable_indexes,
         run_lengths,
         num_runs_out,
         stream,
     )?;
+    let (run_offsets, run_indexes) = temp_storage.split_at_mut(num_cells * 4);
+    let run_offsets = unsafe { run_offsets.transmute_mut() };
```

Comment:
After populating `run_offsets` below, why do we need to keep `run_lengths` around? Threads in `mark_ends_of_runs` and `generate_permutation_matrix_raw` can internally reconstruct `run_length[run_index]` as
`run_offsets[run_index + 1] - run_offsets[run_index]` or
`run_offsets[run_index] - run_offsets[run_index - 1]`
(depending whether the scan is inclusive or exclusive) right?

(I don't think dropping `run_lengths` would help reduce the overall memory high water mark though)